### ISSUE TITLE
General compilation problems

### ISSUE THREAD

**Issue by Chara-Reborn** [crc-32/BrowseNX#1]
Line 15 (buildwin.bat) and Line 16 (build.sh) reference a tool called "npdmtool"
Where would we find this to _actually_ compile BrowseNX?

**Comment by crc-32** [comment]
Ideally, preinstalled by devkitpro (at least on linux) if you have the switch-tools package installed. See: https://github.com/switchbrew/switch-tools
(I'm unsure how well it builds on windows as I just threw together the bat, likely needs fixes)

**Comment by Chara-Reborn** [comment]
I believe I have the package installed.
![image](https://user-images.githubusercontent.com/19896671/51464251-1edf5080-1d33-11e9-995d-ad8587f6fe94.png)

That should be cleared up in that case. But I'm getting errors compiling. I am assuming I'm missing something else as well, not exactly sure what.

Here are my logs when just making the whole thing.
```
C:\Users\Kohai\Desktop\BrowseNX-master>make
main.c
aarch64-none-elf-gcc -MMD -MP -MF /home/Kohai/Desktop/BrowseNX-master/build/main.d -g -Wall -O2 -ffunction-sections -march=armv8-a -mtune=cortex-a57 -mtp=soft -fPIE  -I/home/Kohai/Desktop/BrowseNX-master/include -I/opt/devkitpro/portlibs/switch/include -I/opt/devkitpro/libnx/include -I/home/Kohai/Desktop/BrowseNX-master/build -D__SWITCH__ -c /home/Kohai/Desktop/BrowseNX-master/source/main.c -o main.o
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c: In function 'main':
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:15:5: error: unknown type name 'AppletHolder'; did you mean 'AppletHookFn'?
     AppletHolder aHold;
     ^~~~~~~~~~~~
     AppletHookFn
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:16:5: error: unknown type name 'AppletStorage'; did you mean 'AppletHookFn'?
     AppletStorage aStore;
     ^~~~~~~~~~~~~
     AppletHookFn
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:17:5: error: unknown type name 'LibAppletArgs'
     LibAppletArgs aArgs;
     ^~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:18:5: error: unknown type name 'SwkbdConfig'
     SwkbdConfig kbd;
     ^~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:27:18: warning: implicit declaration of function 'swkbdCreate'; did you mean 'shmemCreate'? [-Wimplicit-function-declaration]
             rc = swkbdCreate(&kbd, 0);
                  ^~~~~~~~~~~
                  shmemCreate
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:29:17: warning: implicit declaration of function 'swkbdConfigMakePresetDefault' [-Wimplicit-function-declaration]
                 swkbdConfigMakePresetDefault(&kbd);
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:30:17: warning: implicit declaration of function 'swkbdConfigSetOkButtonText' [-Wimplicit-function-declaration]
                 swkbdConfigSetOkButtonText(&kbd, "Go");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:31:17: warning: implicit declaration of function 'swkbdConfigSetGuideText' [-Wimplicit-function-declaration]
                 swkbdConfigSetGuideText(&kbd, "Enter a URL");
                 ^~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:32:17: warning: implicit declaration of function 'swkbdConfigSetInitialText' [-Wimplicit-function-declaration]
                 swkbdConfigSetInitialText(&kbd, "https://");
                 ^~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:33:22: warning: implicit declaration of function 'swkbdShow' [-Wimplicit-function-declaration]
                 rc = swkbdShow(&kbd, tmpoutstr, sizeof(tmpoutstr));
                      ^~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:48:13: warning: implicit declaration of function 'swkbdClose'; did you mean 'shmemClose'? [-Wimplicit-function-declaration]
             swkbdClose(&kbd);
             ^~~~~~~~~~
             shmemClose
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:58:5: warning: implicit declaration of function 'appletCreateLibraryApplet' [-Wimplicit-function-declaration]
     appletCreateLibraryApplet(&aHold, AppletId_web, LibAppletMode_AllForeground);
     ^~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:58:39: error: 'AppletId_web' undeclared (first use in this function); did you mean 'AppletType'?
     appletCreateLibraryApplet(&aHold, AppletId_web, LibAppletMode_AllForeground);
                                       ^~~~~~~~~~~~
                                       AppletType
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:58:39: note: each undeclared identifier is reported only once for each function it appears in
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:58:53: error: 'LibAppletMode_AllForeground' undeclared (first use in this function)
     appletCreateLibraryApplet(&aHold, AppletId_web, LibAppletMode_AllForeground);
                                                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:59:5: warning: implicit declaration of function 'libappletArgsCreate' [-Wimplicit-function-declaration]
     libappletArgsCreate(&aArgs, 0x50000);
     ^~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:60:5: warning: implicit declaration of function 'libappletArgsPush' [-Wimplicit-function-declaration]
     libappletArgsPush(&aArgs, &aHold);
     ^~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:61:5: warning: implicit declaration of function 'appletCreateStorage' [-Wimplicit-function-declaration]
     appletCreateStorage(&aStore, 8192);
     ^~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:70:5: warning: implicit declaration of function 'appletStorageWrite'; did you mean 'fsStorageWrite'? [-Wimplicit-function-declaration]
     appletStorageWrite(&aStore, 0, indata, 8192);
     ^~~~~~~~~~~~~~~~~~
     fsStorageWrite
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:71:5: warning: implicit declaration of function 'appletHolderPushInData' [-Wimplicit-function-declaration]
     appletHolderPushInData(&aHold, &aStore);
     ^~~~~~~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:72:5: warning: implicit declaration of function 'appletHolderStart' [-Wimplicit-function-declaration]
     appletHolderStart(&aHold);
     ^~~~~~~~~~~~~~~~~
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:73:5: warning: implicit declaration of function 'appletHolderJoin'; did you mean 'appletHook'? [-Wimplicit-function-declaration]
     appletHolderJoin(&aHold);
     ^~~~~~~~~~~~~~~~
     appletHook
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:74:5: warning: implicit declaration of function 'appletHolderClose'; did you mean 'appletHook'? [-Wimplicit-function-declaration]
     appletHolderClose(&aHold);
     ^~~~~~~~~~~~~~~~~
     appletHook
C:/Users/Kohai/Desktop/BrowseNX-master/source/main.c:75:5: warning: implicit declaration of function 'appletStorageClose'; did you mean 'fsStorageClose'? [-Wimplicit-function-declaration]
     appletStorageClose(&aStore);
     ^~~~~~~~~~~~~~~~~~
     fsStorageClose
make[1]: *** [/opt/devkitpro/devkitA64/base_rules:19: main.o] Error 1
make: *** [Makefile:152: build] Error 2
```

**Comment by crc-32** [comment]
Make sure you have the latest libnx, it requires applet code that was quite recently included in a release, otherwise it's probably something I missed for windows and will need to look into it

**Comment by Chara-Reborn** [comment]
I'm now stuck on actually building the nsp, I've been following the batch script line by line.
This never showed up for me but where is control.nacp supposed to be located?

**Comment by crc-32** [comment]
Try just creating an empty RomFs folder, It actually doesnt use that romfs for anything yet, I'll add a mkdir for that too

**Comment by Chara-Reborn** [comment]
I cleaned up the batch script so it works. You just need npdmtool and hacpack in the root directory.
(As in, where you're running the script from.)

I'll send a pull request with a working script if you can put a copy of npdmtool and hacpack in the repo. (If that makes sense?)

**Comment by crc-32** [comment]
Sure, in the Build-Tools folder

**Comment by crc-32** [comment]
Should be fixed via #3
